### PR TITLE
Search screen and base stats

### LINKED ISSUE #185: Filter cafeteria list by minimum rating

### PR DESCRIPTION
**Pull request by avan1235**
Result of build on [dev server](http://eaterslab-dev.herokuapp.com/download/EatersLabBeta.apk)
Closes #185 (as server alread have review avg implemented on master) and #34 finally

### CODE REVIEW

File: android/app/src/main/java/labs/pooh/eaterslab/repository/dao/AverageDishReviewStatsDao.kt [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+import labs.pooh.eaterslab.client.models.AverageDishReviewStats
+import org.threeten.bp.OffsetDateTime
+
+data class AverageDishReviewStatsDao(
+    var id: Int?,
+    var timestamp: OffsetDateTime,
+    var value: Double
+)
+
+fun AverageDishReviewStats.toDao() = AverageDishReviewStatsDao(id, timestamp!!, value!!.toDouble())
```
Comment (kantoniak) [inline android/app/src/main/java/labs/pooh/eaterslab/repository/dao/AverageDishReviewStatsDao.kt:12]: \n

File: android/app/src/main/java/labs/pooh/eaterslab/repository/dao/OccupancyStatsDao.kt [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+
+import labs.pooh.eaterslab.client.models.OccupancyStats
+
+data class OccupancyStatsDao(
+    var id: Int?,
+    var timestamp: org.threeten.bp.OffsetDateTime,
+    var occupancyRelative: Double
+)
+
+fun OccupancyStats.toDao() = OccupancyStatsDao(id, timestamp!!, occupancyRelative!!.toDouble())
```
Comment (kantoniak) [inline android/app/src/main/java/labs/pooh/eaterslab/repository/dao/OccupancyStatsDao.kt:11]: \n

File: android/app/src/main/java/labs/pooh/eaterslab/ui/activity/search/SearchViewModel.kt [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+                minAvgReview = avgReview
+            )
+            cafeterias?.forEach { cafeteria ->
+                    cafeteria.downloadContent()
+                    addSearchedCafeteria(cafeteria)
+            }
+            onFinished()
+        }
+    }
+}
```
Comment (kantoniak) [inline android/app/src/main/java/labs/pooh/eaterslab/ui/activity/search/SearchViewModel.kt:51]: \n

File: android/app/src/main/java/labs/pooh/eaterslab/ui/fragment/dialogs/TimePickerDialogFragment.kt [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+class TimePickerDialogFragment(private val holder: PickedTimeHolder) : DialogFragment() {
+
+    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
+        val timeNow = Calendar.getInstance()
+        return TimePickerDialog(activity, { _, hour, minute ->
+            holder.setTime(hour, minute)
+        }, timeNow.get(Calendar.HOUR_OF_DAY), timeNow.get(Calendar.MINUTE), true)
+    }
+}
```
Comment (kantoniak) [inline android/app/src/main/java/labs/pooh/eaterslab/ui/fragment/dialogs/TimePickerDialogFragment.kt:18]: \n

File: android/app/src/main/java/labs/pooh/eaterslab/ui/fragment/stats/StatsViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        val minTime = now.minusMonths(1).withHour(minHour).zeroMinutes()
+        val data = repository.cafeteriaReviewStatsRead(cafeteriaId, minTime, now, Int.MAX_VALUE, StatsGroupOption.BY_DAY)
+        _currentMonthAverageReviewData.value = data
+    }
+}
+
+private fun OffsetDateTime.zeroMinutes() = this
+    .withMinute(0)
+    .withSecond(0)
+    .withNano(0)
```
Comment (kantoniak) [inline android/app/src/main/java/labs/pooh/eaterslab/ui/fragment/stats/StatsViewModel.kt:62]: \n

File: android/app/src/main/java/labs/pooh/eaterslab/ui/view/PickedTimeHolder.kt [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    fun toTimeApiFormat(): TimeApi? {
+        val currHour = hour
+        val currMinute = minute
+        return if (currHour == null || currMinute == null) {
+            null
+        } else {
+            TimeApi(currHour, currMinute)
+        }
+    }
+}
```
Comment (kantoniak) [inline android/app/src/main/java/labs/pooh/eaterslab/ui/view/PickedTimeHolder.kt:38]: \n

File: android/app/src/main/java/labs/pooh/eaterslab/util/ActivityUtils.kt [hunk 7]
```diff
@@ -14,3 +15,9 @@ inline fun <reified T : Activity> Context.start(intentBlock: Intent.() -> Unit =
 
 fun Activity.snack(message: String, duration: Int = Snackbar.LENGTH_SHORT)
         = Snackbar.make(findViewById(android.R.id.content), message, duration).show()
+
+inline fun withVisible(view: View, block: (View) -> Unit) {
+    view.visibility = View.VISIBLE
+    block(view)
+    view.visibility = View.INVISIBLE
+}
```
Comment (kantoniak) [inline android/app/src/main/java/labs/pooh/eaterslab/util/ActivityUtils.kt:23]: \n

File: android/app/src/main/java/labs/pooh/eaterslab/util/PlotUtils.kt [hunk 8]
```diff
@@ -177,5 +175,71 @@ open class DiscreteLinePlot(context: Context) : AbstractStandardPlot(context) {
...
+        val index = ticksIndexer(domainValue)
+        val ticks = ticksValues.size
+        if (ticks < 2) {
+            return widthWithPadding / 2
+        }
+
+        val distance = (widthWithPadding - 2 * padding) / (ticks - 1)
+        return index * distance + padding
+    }
+}
```
Comment (kantoniak) [inline android/app/src/main/java/labs/pooh/eaterslab/util/PlotUtils.kt:245]: \n

File: android/app/src/main/res/drawable/edit_text_search.xml [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+        <shape android:shape="rectangle">
+            <solid android:color="?backgroundLightColor"/>
+            <stroke
+                android:color="?backgroundLightColor"
+                android:width="1dp"
+                />
+            <corners android:radius="16dp"/>
+        </shape>
+    </item>
+</selector>
```
Comment (kantoniak) [inline android/app/src/main/res/drawable/edit_text_search.xml:13]: \n

File: android/app/src/main/java/labs/pooh/eaterslab/repository/dao/AverageDishReviewStatsDao.kt (hunk 1 above)
Comment (avan1235) [reply android/app/src/main/java/labs/pooh/eaterslab/repository/dao/AverageDishReviewStatsDao.kt:12]: fixed

File: android/app/src/main/java/labs/pooh/eaterslab/ui/activity/search/SearchViewModel.kt (hunk 3 above)
Comment (avan1235) [reply android/app/src/main/java/labs/pooh/eaterslab/ui/activity/search/SearchViewModel.kt:51]: fixed

File: android/app/src/main/java/labs/pooh/eaterslab/repository/dao/OccupancyStatsDao.kt (hunk 2 above)
Comment (avan1235) [reply android/app/src/main/java/labs/pooh/eaterslab/repository/dao/OccupancyStatsDao.kt:11]: fixed

File: android/app/src/main/java/labs/pooh/eaterslab/ui/fragment/dialogs/TimePickerDialogFragment.kt (hunk 4 above)
Comment (avan1235) [reply android/app/src/main/java/labs/pooh/eaterslab/ui/fragment/dialogs/TimePickerDialogFragment.kt:18]: fixed

File: android/app/src/main/java/labs/pooh/eaterslab/ui/fragment/stats/StatsViewModel.kt (hunk 5 above)
Comment (avan1235) [reply android/app/src/main/java/labs/pooh/eaterslab/ui/fragment/stats/StatsViewModel.kt:62]: fixed

File: android/app/src/main/java/labs/pooh/eaterslab/ui/view/PickedTimeHolder.kt (hunk 6 above)
Comment (avan1235) [reply android/app/src/main/java/labs/pooh/eaterslab/ui/view/PickedTimeHolder.kt:38]: fixed

File: android/app/src/main/java/labs/pooh/eaterslab/util/ActivityUtils.kt (hunk 7 above)
Comment (avan1235) [reply android/app/src/main/java/labs/pooh/eaterslab/util/ActivityUtils.kt:23]: fixed

File: android/app/src/main/java/labs/pooh/eaterslab/util/PlotUtils.kt (hunk 8 above)
Comment (avan1235) [reply android/app/src/main/java/labs/pooh/eaterslab/util/PlotUtils.kt:245]: fixed

File: android/app/src/main/res/drawable/edit_text_search.xml (hunk 9 above)
Comment (avan1235) [reply android/app/src/main/res/drawable/edit_text_search.xml:13]: fixed

**Comment by kantoniak** [comment]
Is it ready for review?

**Comment by avan1235** [comment]
> Is it ready for review?

@kantoniak  Hope it is

**Pull request merged**